python/snewpy/rate_calculator.py

```diff
@@ -13,11 +13,281 @@
+            Can be a function of E_nu, an array (of size `len(bins_true)-1`) or a scalar.
+        sigma: Callable or u.Quantity[energy]
+            Sigma of the smeared energy distribution.
+            Can be a function of E_nu, an array (of size `len(bins_true)-1`) or a scalar.
+
+        Examples
+        --------
+        
+        >>> e_true =np.linspace(0,30,501)<<u.MeV
+        >>> e_smear=np.linspace(0,30,501)<<u.MeV
+        >>> #IBD process with minimal smearing, but with constant shift E_det = E_nu-0.87 MeV
+        >>> smear_IBD = SmearingMatrix.from_Gaussian(e_true,e_smear, mean=lambda e_nu:e_nu-0.87*u.MeV)
+
+        >>> #NC on C12 process with large smearing and constant energy response at 15.1 MeV
+        >>> smear_NC_C = SmearingMatrix.from_Gaussian(e_true,e_smear, mean=15.1*u.MeV, sigma=2*u.MeV)
+
+        """
+        e_t = center(bins_true)<<u.MeV
+        e_s = bins_smeared<<u.MeV
+        
+        if callable(mean):
+            mean = mean(e_t)
+        if callable(sigma):
+            sigma = sigma(center(bins_true))
+        #expand the scalar to required size
+        mean = np.ones(e_t.shape)*mean
+        sigma = np.ones(e_t.shape)*sigma
+        #true energy will be axis=0    
+        mean = np.atleast_2d(mean).T<<u.MeV
+        sigma = np.atleast_2d(sigma).T<<u.MeV
+        #smeared energy will be axis=1
+        e_s = np.atleast_2d(e_s)
+        distr =  st.norm(loc=mean, scale=sigma)
+        #calculate integral in each bin
+        cdf = distr.cdf(e_s)
+        pdf = np.diff(cdf, axis=1)
+        return cls(bins_true, bins_smeared, pdf)
+
+class FunctionOfEnergy:
+    """A wrapper around a given function of energy, for example interaction cross-section, or a detection efficiency.
+    This helper class will perform a correct operation when multiplied to a :class:`Container` object - rate or flux.
+    """
+    def __init__(self, callable):
+        """
+        Parameters
+        ----------
+        callable: Callable
+            A function of one parameter (energy). This can be an analitical function, or an interpolation of a (E,value) dataset
+        """
+        self.value = callable
+    def __mul__(self, f:Container)->Container:
+        e = f.energy #Define sample points
+        if not f.can_integrate('energy'): #we have bins, let's use central values for sampling
+            e = center(f.energy)
+        return f*self.value(e)
+    def __call__(self, energy):
+        return self.value(energy)
+        
+    @classmethod
+    def from_threshold(cls, e_min=1<<u.MeV):
+        return cls(lambda e: 1*(e>e_min))
+
+class DetectionChannel:
+    """Description of a single detection channel in the detector"""
+    
+    def __init__(self, flavor:Flavor, xsec:callable, smearing:SmearingMatrix=None,
+                 efficiency:FunctionOfEnergy=1., weight:float=1.):
+        """
+        Parameters
+        ----------
+        name:str
+            channel name
+        flavor:Flavor or list[Flavor]
```

Comment:
I am not 100% convinced that one DetectionChannel is defined by one flavor, that's the case of IBD but not of scattering/NC channels, where all flavors are involved. I guess in that case if I want the total for (e.g. pES), I should just define four detection channels (nue, anti-nue, nu_x, anti-nux) and sum them after right?

Reply:
Oops, the dosctring is wrong, thanks for pointing it out!
I agree, it'd be better to have a single channel with several flavors. But this DetectionChannel class was made to represent 1to1 the channels in the SNOwGLoBES, and they are separate for each neutrino flavor:
```
nc_nue_C12
nc_nuebar_C12
nc_numu_C12
nc_numubar_C12
```

We can implement this (many flavors in one channel) as an additional feature, if needed, in a separate PR.